proposals/1859_github-auth/README.md

```diff
@@ -0,0 +1,500 @@
+  # CLI-based auth, but still allows browser-based auth.
+  allowPersonalAccessTokens:
+    # Allow or disallow the use of GitHub Fine-grained PATs
+    # to authenticate to Kubernetes clusters. If they
+    # are disallowed for your GitHub org, then also
+    # disallow them here so your users can get a nice
+    # error message when trying to use a fine-grained PAT.
+    # Defaults to false.
+    fineGrained: true
+    # Allow or disallow the use of GitHub Classic PATs
+    # to authenticate to Kubernetes clusters. If they
+    # are disallowed for your GitHub org, then also
+    # disallow them here so your users can get a nice
+    # error message when trying to use a classic PAT.
+    # Defaults to false.
+    classic: false
+  client:
+    # The name of Secret in the same namespace which holds the
+    # client ID and client secret of a GitHub App or OAuth App.
+    # This will only be used for web browser-based auth flows.
+    # Required.
+    secretName: github-client-credentials
+status:
+  conditions: # conditions TBD. Shows validations, if any.
+  phase: # ready or error, if there are conditions.
+```
+
+#### Web Browser-based Authentication using GitHubIdentityProvider
+
+The client ID and client secret configured on the GitHubIdentityProvider can be for either a GitHub App or an OAuth App.
+This flexibility can be made possible because there are many similarities in the OAuth 2.0 authorization flow used by
+both. To the extent that we can ignore the differences, Pinniped will not need to know if it is acting as a GitHub App
+or an OAuth App.
+
+The authorization flow for GitHub apps is described in
+[this doc](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-a-user-access-token-for-a-github-app)
+and the flow for OAuth Apps is described in
+[this doc](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps).
+
+From a client's point of view, the main difference is that access tokens for a GitHub App may expire, depending how the
+app was configured on GitHub. However, if the access token expires, then it is always after 8 hours. That duration is
+not currently configurable. This gives Pinniped a nice option to save implementation time and complexity: we can simply
+ignore GitHub's OAuth refresh flow. The user's Pinniped session should expire either when their GitHub access token
+stops working (i.e. was expired or revoked) or whenever Pinniped's hardcoded 9-hour session limit is reached.
+If it is an expiring access token, then the session will end after 8 hours due to the token expiration.
+If it is not an expiring access token, then the session will end after 9 hours. That's close enough to Pinniped's
+concept of "once-per 9 hours auth" that it's not worth the extra effort to implement the refresh flow with GitHub.
+
+The other difference from a client's point of view is that GitHub Apps do not use scopes. However, the `scope` param
```

Comment:
Does this mean "Github App" vs any "Oauth App" or a Github OAuth App?

Reply:
Updated all usages of "OAuth App" to be "GitHub OAuth App" to clarify this.